src/lib/actor/actor.ml

```diff
@@ -0,0 +1,260 @@
+open Async
+open Core_kernel
+
+type 'state msg_processed = MNext of 'state | MExit | MUnprocessed
+
+type ('state, 'response) req_processed =
+  | RNext of ('state * 'response)
+  | RExit of 'response
+  | RUnprocessed
+
+type (_, _, _, _) overflow_behavior =
+  | Throw : ('state, 'data, unit Or_error.t, [ `Throw ]) overflow_behavior
+  | Drop_head :
+      [ `Warns | `No_warns ]
+      -> ('state, 'data, unit, [ `Drop_head ]) overflow_behavior
+  | Drop_and_call_head :
+      ('state -> 'data -> 'returns)
+      -> ( 'state
+         , 'data
+         , 'returns option
+         , [ `Drop_and_call_head ] )
+         overflow_behavior
+  | Push_back
+      : ('state, 'data, unit Deferred.t, [ `Push_back ]) overflow_behavior
+
+type (_, _, _, _) channel_type =
+  | Infinity : ('state, 'data, unit, [ `Infinity ]) channel_type
+  | With_capacity :
+      [ `Capacity of int ]
+      * [ `Overflow of ('state, 'data, 'returns, 'behavior) overflow_behavior ]
+      -> ('state, 'data, 'returns, 'behavior) channel_type
+
+module DummyMessage = struct
+  type t = unit [@@deriving to_yojson]
+
+  let handler ~state ~message:() = Deferred.return (MNext state)
+end
+
+module DummyRequest = struct
+  type _ t = Nothing : unit t
+
+  let handler :
+      type response.
+         state:'state
+      -> request:response t
+      -> ('state, response) req_processed Deferred.t =
+   fun ~state ~request:Nothing -> Deferred.return (RNext (state, ()))
+end
+
+module WithRequest (DataMessage : sig
+  type t
+
+  val to_yojson : t -> Yojson.Safe.t
+end) (Request : sig
+  type _ t
+end) =
+struct
+  exception
+    ActorDataInboxOverflow of
+      { name : Yojson.Safe.t; capacity : int; attempt_enqueing : DataMessage.t }
+
+  exception RunningActorSpawned of { name : Yojson.Safe.t }
+
+  type 'state request_handler =
+    { f :
```

Comment:
This has to be a struct instead of a function because locally abstracted function can't be passed as params directly.